tests/apollo/test_skvbc_backup_restore.py

```diff
@@ -183,7 +183,112 @@ async def test_checkpoint_propagation_after_restarting_all_replicas_under_load(s
-        self.assertGreaterEqual(next_checkpoint_before + 1, next_checkpoint_after)
+        self.assertGreaterEqual(next_checkpoint_after, next_checkpoint_before + 1)
+
+    @with_trio
+    @with_bft_network(start_replica_cmd_with_vc_timeout("20000"), selected_configs=lambda n, f, c: n == 7)
+    @with_constant_load
+    async def test_checkpoint_propagation_after_restarting_majority_replicas_under_load(self, bft_network, skvbc,
+                                                                                        nursery):
+        """
+        Here we trigger a checkpoint, restart all replicas in a random order with a delay in-between,
+        both while stopping and starting. We verify checkpoint persisted upon restart and then trigger
+        another checkpoint. We make sure checkpoint is propagated to all the replicas.
+        1) Given a BFT network, we start n - f replicas
+        2) Send sufficient number of client requests to trigger checkpoint protocol
+        3) Stop n - f replicas in a random order (with 10/n seconds delay in between)
+        4) Start all replicas in a random order (with 10 seconds delay in between)
+        5) Wait for view to stabilize
+        6) Wait for state transfer to stop
+        7) Send sufficient number of client requests to trigger another checkpoint
+        8) Make sure checkpoint propagates to all the replicas
+        """
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+
+        n = bft_network.config.n
+        f = bft_network.config.f
+
+        initial_primary = await bft_network.get_current_primary()
+
+        # choose a random set of f replicas to stop
+        f_replicas_stopped_early = bft_network.random_set_of_replicas(f, without={initial_primary})
+        bft_network.stop_replicas(f_replicas_stopped_early)
+
+        replicas_up = [i for i in bft_network.all_replicas() if i not in f_replicas_stopped_early]
+        self.assertEqual(len(replicas_up), n - f, "Make sure n-f replicas are up.")
+
+        initial_view = 0
+
+        # trigger a checkpoint and verify checkpoint propagation among all live replicas
+        checkpoint_before = await bft_network.wait_for_checkpoint(replica_id=initial_primary)
+
+        await self._fill_and_wait_for_checkpoint_under_constant_load(
+            skvbc,
+            bft_network,
+            initial_nodes=bft_network.all_replicas(without=f_replicas_stopped_early),
+            num_of_checkpoints_to_add=1,
+            verify_checkpoint_persistency=False,
+            assert_state_transfer_not_started=False)
+
+        checkpoint_after = await bft_network.wait_for_checkpoint(replica_id=initial_primary)
+        self.assertGreaterEqual(checkpoint_after, checkpoint_before + 1)
+
+        # stop n replicas in a random order with a delay of 10/n seconds in between, so that
+        # all replicas are stopped by 10 seconds, and no view change is triggered during stopping
+        # of replicas.
```

Comment:
Why avoid the view change while stopping the replicas? In a real-world scenario of backup-restore, we can't guarantee that the primary won't be stopped before the others, or the length of the delays.

Reply:
Yes, that makes sense.